### ISSUE TITLE
Overhaul of allocation, reservation and preemption logic

### ISSUE THREAD

**Issue by wangdatan** [apache/incubator-yunikorn-core#74]
**Why this issue?**

- The previous approach for allocation and preemption: 

```
Allocation --------------------------->
Preemption -> ComputeIdealAllocation ->  SingleStepSchedule. 
  
  SingleStepSchedule:
     -> FindAllocationAsks (which find a list of allocation requests)
     -> TryBatchAllocation (and pass in the allocation requests).
        -> allocate:
           -> (If it is preemption): CrossQueuePreemptionAllocate
             -> trySurgicalPreemptionOnNode 
           -> (If it is non-preemption): RegularAllocate. 

        -> Then use a complex algorithm [1] to blacklist requests 
```

What does the algorithm [1] looks like? 

```
    for failedAllocationKey := range failedToAllocationKeys {
        if preemptionParam.crossQueuePreemption {
            preemptionParam.blacklistedRequest[failedAllocationKey] = true
        } else {
            curWaitValue := m.waitTillNextTry[failedAllocationKey]
            if curWaitValue == 0 {
                curWaitValue = 2
            } else if curWaitValue < (1 << 20) {
                // Increase missed value if it is less than 2^20 (TODO need do some experiments about this value)
                curWaitValue <<= 1
            }
            m.waitTillNextTry[failedAllocationKey] = curWaitValue
        }
    }
```

After thought, I don't think it is correct, let's say there're 3 requests in a queue sorted by priority is R1 > R2 > R3.

```
1st time: Allocate order: R1 > R2 > R3. 
Assume, there's no resource available on the cluster to satisfy R1, R2.
So scheduler allocate 1 space for R3, and waitValue (see Algorithm [1]) of R1, R2 became 2 

2nd time: Allocation for R1 and R2 will be skipped, scheduler directly goes to R3 without looking at R1 and R2.
```

**Proposed Solution:**

Go to similar solution of YARN CapacityScheduler/FairScheduler introduce resource reservation on node to preserve allocation order. 

New function invoking structure considering allocation and preemption. 

```
SingleStepAllocation: 
   -> TryAllocate()

      for q in ordered-queues: 
         for app in ordered-apps:
            for req in app.requests: 
               if allocate(req) != nil: // allocated or reserved something 
                  break-the-nested-for-loop 

allocate(req): 
   // Very similar to existing regularAllocate, 
   // but reserve req on node if cannot allocate 
   // because of available space.


Preemption:
   -> TryPreemption() 

      calculateIdealAllocation

      for req in reservedRequests: 
        if overAllocated(req):
          drop-the-req-reservation
        else:
          try to preemption for req;
          // Very similar to trySurgicalPreemptionOnNode
```

**Comment by wangdatan** [comment]
cc: @yangwwei, @sunilgovind, @wilfred-s, @TaoYang526

**Comment by wangdatan** [comment]
This proposal will invalidate the batch allocation procedure, which will require allocate one request at a time. (But with proper locking, we can still support multi-thread allocation proposal like how YARN CapacityScheduler global scheduling works). 

This should not impact gang scheduling for the future since allocation-proposal can still include multiple requests. 

This could have some impact on the node sorting algorithm, since @yangwwei mentioned there are some works to only sort node once per patch. (That is my understanding per my previous conversation with Weiwei, but I didn't see the behavior after apply the PR #73). If #73 is the final behavior, the proposed change will not impact performance.

I'm already started to work on a patch on top of #73.

**Comment by wangdatan** [comment]
Found a comment in the code which might be helpful to understand algorithm [1]: 

```
    // Wait till next try
    // (It is designed to be accessed under a single goroutine, no lock needed).
    // This field has dual purposes:
    // 1) When a request is picky, scheduler will increase this value for every failed retry. So we don't need to
    //    look at picky-requests every time.
    // 2) This also indicate starved requests so preemption can do surgical preemption based on this.
    waitTillNextTry map[string]uint64
```

**Comment by yangwwei** [comment]
Thanks @wangdatan for working on this!

> This could have some impact on the node sorting algorithm, since @yangwwei mentioned there are some works to only sort node once per batch.

That is not part of #73. That is supposed to be added after #73 is merged.

For the reservation, I have few questions

1) Since we can scan all qualified nodes for allocation, the interesting problem is when there is a R outstanding, what's the algorithm to find out the node for reservation? 
2) The proposed solution doesn't include the info about when to unreserve, do we do it each time before allocating any new containers, as well as each time there is any released container?
3) If R is reserved on node NodeA, a container is released on NodeB and now has space for R, do we unreserve R from NodeA and allocate it on NodeB? Or it has to wait for NodeA resource?
4) Do we allow to reserve a container multiple times?

**Comment by TaoYang526** [comment]
Thanks @wangdatan for working on this.
@yangwwei asked some questions above, which are also what I want to know.
For reservation logic in YARN CapacityScheduler, I'm afraid it's not applicable for global scheduling now and have created YARN-9598 for that, @yangwwei have told me that there was a plan to refactor the reservation logic so we didn't keep tracking that issue. 
I would like to know some details to make sure same problems won't happen in yunikorn. Thanks.

**Comment by wilfred-s** [comment]
In my testing for #73  I have found another major issue in the preemption code: preemption is not tracked against an application or even a candidate allocation. This means that the pre-emption that is being worked on and one that is in progress for a different allocation, of the same application, cannot be distinguished.
This could in turn lead to the scheduler thinking that the node has space and thus not need to preempt while there is no space and preemption should be performed. The allocation then fails as it would push the node over its allocation limit.

I am not sure if there are other gaps in the current code.

Looking at how and were we track preemption we need to do it differently. I don't think using a third set of objects, like the preemption queues etc is going to work.

**Comment by wangdatan** [comment]
Thanks @yangwwei 
> 
> That is not part of #73. That is supposed to be added after #73 is merged.

Can we hold that a bit since to me. getting correctness of the allocation will be more important? We can optimize sort algorithm post this patch to avoid additional changes and more complexities for the overhual.

> 
> For the reservation, I have few questions
> 
>     1. Since we can scan all qualified nodes for allocation, the interesting problem is when there is a R outstanding, what's the algorithm to find out the node for reservation?

Scheduler will perform similar to YARN: scheduler will pick up head-of-line request, and reserve it on one of the node. Algorithm could be as simple as a greedy algorithm: find a node which has closest avail resource for a given request.

> 
>     2. The proposed solution doesn't include the info about when to unreserve, do we do it each time before allocating any new containers, as well as each time there is any released container?
> 
>     3. If R is reserved on node NodeA, a container is released on NodeB and now has space for R, do we unreserve R from NodeA and allocate it on NodeB? Or it has to wait for NodeA resource?

To me 2/3 are the same question: to me, the reservation is just a placeholder, the scheduler should check other nodes if the scheduler can immediately move reservation to another node to make allocation happen. 

> 
>     4. Do we allow to reserve a container multiple times?

I would prefer no to avoid exact complexities, but we can actively look at reserved containers and allocate them.

**Comment by wangdatan** [comment]
@TaoYang526 

> For reservation logic in YARN CapacityScheduler, I'm afraid it's not applicable for global scheduling now and have created YARN-9598 for that, @yangwwei have told me that there was a plan to refactor the reservation logic so we didn't keep tracking that issue.
> I would like to know some details to make sure same problems won't happen in yunikorn. Thanks.

Agree, I think we should be able to avoid the issue. The reason for YARN CS is, we have to support  node-heartbeat allocation and global scheduling (it might be a mistake which we support both). This makes reservation logic in CS is overcomplicated which have lots of trouble to support globalscheduling + reservation. 

The YK is global-scheduling only mode, so I don't see an issue which we cannot support it.

The approach in my mind is: 

```
During allocation:

SingleSchedule:
  - First check reserved requests (sorted by age of reservation): 
    for 
  - 
```

**Comment by wangdatan** [comment]
> In my testing for #73 I have found another major issue in the preemption code: preemption is not tracked against an application or even a candidate allocation. This means that the pre-emption that is being worked on and one that is in progress for a different allocation, of the same application, cannot be distinguished.

Sorry I don't fully understand this, could you elaborate about this? 

> Looking at how and were we track preemption we need to do it differently. I don't think using a third set of objects, like the preemption queues etc is going to work.

I think it is fine to merge #73 if it breaks preemption, we should fix preemption logic separately (may not even in this PR).

**Comment by wilfred-s** [comment]
To explain further what I see:
* Application A is started in a cluster and runs in QueueA.
* The cluster has 2 nodes N1 and N2. The application is large enough that it fully occupies the two nodes in the cluster.
* A second application B is started which runs in QueueB after the first app A has taken all the resources available.
* Both queues have the same configuration set and preemption is turned on.
* both applications ask for the same sized allocations.

What happens in failure case 1:
1. To allocate resources for app B the scheduler starts preempting allocations from app A. After selecting an allocation from app A the resources on the node are marked as under preemption. The scheduler sends of the request to release the allocation from app A to the cache and the new allocation from app B to take over.
1. At this point the scheduler starts working on a second allocation step as app B is still under its share. 
1. The allocation from A is released on a node.
1. The second allocation round for app B checks if there are resources available on a node and sees that there are some resources available as the allocation from app A is released.
At this point the scheduler decides not to preempt (enough resources available) and approves the allocation for app B and sends it to the cache.
1. The first allocation for app B then is confirmed and processed. This brings the node back to its limit.
1. The second allocation for app B then starts processing. This processing fails as the node is already back at its limit again.

Failure case number 2:
Step 1 and 2 are the same as above.
3. The second allocation round for app B checks if there are resources available on a node. This happens before the previously selected allocation for preemption is released. The scheduler might again select the same allocation for the same node to preempt.
3. The allocation from the first round of preemption is processed and released.
3. app B is allocated the preempted resource (as per allocation round 1)
3. The allocation from the second round of preemption is processed, release fails as it is already released.
3. app B second round allocation fails as the resource on the node is not available.

This is caused by the fact that the preemption is just linked to resources and not to an allocation. We also do not know if the resources that are under preemption will be fully used by the allocation that is linked to it. We can only see the net result of the preemption and the in progress allocation.

So unless we change the way we do this preemption will keep giving strange results.

**Comment by TaoYang526** [comment]
Thanks @wangdatan for the clarification.
I think there maybe lots of differences for reservation logic between YK and YARN, for example, R1 (priority=1) is reserved on some nodes and later R2 with higher priority(priority=2) comes, I think we should guarantee that R2 can preempt reservation for R1 if it can't be allocated on unreserved nodes, since most apps  on K8s are long-running and higher-priority requests can't be blocked too long. Right? If yes, it seems that inter-queue fairness, inner-queue priorities should be considered together with reservation logic.

**Comment by wangdatan** [comment]
> To explain further what I see:
> 

Thanks @wilfred-s , will take it into account when start the implementation of preemption.

**Comment by wangdatan** [comment]
>  I think there maybe lots of differences for reservation logic between YK and YARN, for example, R1 (priority=1) is reserved on some nodes and later R2 with higher priority(priority=2) comes, I think we should guarantee that R2 can preempt reservation for R1 if it can't be allocated on unreserved nodes, since most apps on K8s are long-running and higher-priority requests can't be blocked too long. Right? If yes, it seems that inter-queue fairness, inner-queue priorities should be considered together with reservation logic.

Make sense, my idea is to actively drop reservation during preemption. If we do see a need to preempt from a queue. We can aggressively drop reservation and make rooms for new reservations. The reason we can aggressively do that is because reservation is not visible from outside.

**Comment by wangdatan** [comment]
I'm doing the patch now: 

I just finished changes to move methods into places and dropped existing preemption logics, as well as the resource request black-list problematic approach mentioned by the title. I haven't done any reservation and preemption overhaul yet. 

Please feel free to check the implementation in: 

https://github.com/wangdatan/yunikorn-core/commit/fd1e4ee3eb197c9e693247d1e88669f900420234

It looks like there's huge performance impact of my patch (after my patch throughput is 50 pods/sec), comparing to 300 pods/sec before. I still need to spend some time to investigate.

But it will be helpful to get some feedback back refactorings so far (other than preemption logic changes, the existing patch doesn't have any functionality change.

**Comment by wangdatan** [comment]
Ah, I found the root cause is my patch removed batch allocation, now the performance after my change is almost as same as to batch = 1 case before my patch.

Will try to look at how to optimize to bring the perf back.

**Comment by wangdatan** [comment]
Trying to figure out the issue of performance, found two issues: 

1) Node schedulable access has some minor impact of performance: https://github.com/wangdatan/yunikorn-core/commit/ea25a18a513d9d3557e76745a59431450b4a6c8f can bring throughput from ~60 to ~90 per sec. 

2) #76 will be a major issue, after temporarily change the behavior, it can go from ~90 to ~480. I think we should fix the issue separately (also to me existing behavior is confusing).

**Comment by wilfred-s** [comment]
this is implemented
resolved by #81
